### DISCUSSION TITLE
Data are not refreshed after affiliation change in cart

### DISCUSSION THREAD

**Discussion by ahgamal-ms** [microsoft/Dynamics365Commerce.Solutions#33] [Q&A]
in POS We need to exclude some Affiliation on Cart by extension, we have tried to use the OnExecuted trigger for SaveCartRequest to modify the selected Affiliation on cart. the affiliation get changed in cart but the cart discount did not get calculated.

So how can we recalculate cart discounts or force cart recalculation throw extensions or if we should use other trigger to change the affiliation.

example is: 
•	Customer has attached to his master data affiliations EMPLOYEE and MKT008
•	In the setup table the affiliations to be excluded are MKT001, MKT002, MKT008
•	Result in the cart should be that we have only the affiliation EMPLOYEE, because MKT008 coming from the customer has been excluded
 
We think the solution could consists of forcing a refresh on the cart after the custom code is run within “OnExecuted” method, but we haven’t found yet a way to recalculate the discounts.

**Comment by madyke** [comment]
Hi @ahgamal-ms - I'm not familar with how the discount calculations work, but I see a couple of requests in the API that may be useful, you could give these a try to see if they accomplish what you're looking for:
- `TotalDiscountAmountOperationRequest`
- `RefreshCartClientRequest`
- `CalculateTotalOperationRequest`
- `LineDiscountAmountOperationRequest`

If none of those work for your scenario, if you can provide the code you're using now for the `SaveCartRequest` then I can take a look on my dev VM to see if there's another workaround we can recommend.

**Reply by vmarano** [reply]
Hi @madyke - I'm inside the CRT extension, specifically I have implemented the "IRequestTriggerAsync" interface and I can't use the "Requests" you wrote.

Including also @ahgamal-ms


Below the implemented code:

public async Task OnExecuted(Request request, Response response)
            {
                
                if (request.GetType() == typeof(SaveCartRequest))
                {
                    SaveCartRequest req = (SaveCartRequest)request;
                    SaveCartResponse resp = (SaveCartResponse)response;
                    Cart cart;

                    //if (!resp.Cart.CustomerId != "")
                    //{   //JL begin
                    if (request.RequestContext.Runtime.Configuration.IsMasterDatabaseConnectionString)
                    {
                        //JL end
                        cart = ((SaveCartResponse)response).Cart;

                        if (cart.CustomerId != null && cart.CustomerId != "")
                        {
                            IList<AffiliationLoyaltyTier> AffiliationLoyaltyTierList = cart.AffiliationLines;
                            IList<AffiliationLoyaltyTier> AffiliationLoyaltyTierListChecked = new List<AffiliationLoyaltyTier>();
                            List<long> affiliationLoyaltyTierRecid = new List<long>();
                            string resValue;
                            if (!AffiliationLoyaltyTierList.IsNullOrEmpty())
                            {
                                //vmarano - start
                                List<Int64> excludedAffiliations;
                                excludedAffiliations = this.getExcludedAffiliations(request.RequestContext, cart.LoyaltyCardId);
                                //vmarano
                                foreach (AffiliationLoyaltyTier affiliationLoyaltyTier in AffiliationLoyaltyTierList)
                                {
                                    if (!excludedAffiliations.Contains(affiliationLoyaltyTier.AffiliationId))
                                    {
                                        affiliationLoyaltyTierRecid.Add(affiliationLoyaltyTier.AffiliationId);
                                        TransactionServiceClient transactionService = new TransactionServiceClient(req.RequestContext);
                                        ReadOnlyCollection<object> results = transactionService.InvokeExtensionMethod("getCapAmountForAffiliations", cart.CustomerId, cart.ChannelId, 0, affiliationLoyaltyTier.AffiliationId);
                                        resValue = (string)results[0];
                                        if (resValue != "")
                                        {
                                            AffiliationLoyaltyTierListChecked.Add(affiliationLoyaltyTier);
                                        }
                                    }
                                }
                                cart.AffiliationLines = AffiliationLoyaltyTierListChecked;
                            }

                        }
                    }
                }
                //response.SetProperty() .SetProperty("Carts",cartsmodified);
                //}
                await Task.CompletedTask;

            }
